File: assets/scripts/weaponBase.ts

import { _decorator, Component, Vec3, Prefab, instantiate, Node, NodePool } from 'cc';
import { Maze_GlobalMouseListener } from './globalMouseListener';
import { Maze_EasyReference } from './easyReference';
import { Maze_BulletBase } from './bulletBase';
import { Maze_Observer } from './observer/observer';
import std from './thirdparty/tstl/src';
const { property } = _decorator;

export namespace Maze_WeaponBase
{ 
    export abstract class WeaponBase extends Component 
    {
        private static _bulletPoolMap:std.TreeMap<String, NodePool> = new std.TreeMap<String, NodePool>();

        @property (Prefab)
        bulletPrefab:Prefab|null = null;

        @property
        fireRate:number = 1;

        @property
        bulletTimeAlive:number = 2;

        @property
        bulletSpeed:number = 2;

        @property
        damage:number = 5;

        @property
        bulletCollisionGroup:number = 0;

        protected easyReference:Maze_EasyReference.EasyReference|null = null;

        private _fireOn:boolean = false;

        private _fireAllowed:boolean = true;
        public get fireAllowed() : boolean
        {
            return this._fireAllowed;
        }

        private _direction:Vec3 = new Vec3();
        public set direction(val:Vec3)
        {
            this._direction = val;
        }
        public get direction() : Vec3
        {
            return this._direction;
        }

        private _destroyBulletObserver:Maze_Observer.Observer<Maze_BulletBase.DestroyBulletContext> 
        = new Maze_Observer.Observer<Maze_BulletBase.DestroyBulletContext>();

        protected createBullet() : Node|null
        {
            var bulletInstance:Node|null = null;

            if(null != this.bulletPrefab)
            {
                var nodePool:NodePool|null = null;

                if(WeaponBase._bulletPoolMap.has(this.bulletPrefab.name))
                {
                    nodePool = WeaponBase._bulletPoolMap.get(this.bulletPrefab.name);
                }
                else
                {
                    nodePool = new NodePool();
                    WeaponBase._bulletPoolMap.set(this.bulletPrefab.name,nodePool);
                }

                if(0 != nodePool.size())
                {
                    bulletInstance = nodePool.get();

                    if(null != bulletInstance)
                    {
                        var bulletComponent = bulletInstance.getComponent( Maze_BulletBase.BulletBase );
        
                        if(null != bulletComponent)
                        {
                            bulletComponent.reuse();
                        }
                    }
                    else
                    {
                        throw("Error! bulletInstance == null!");
                    }
                }
                else
                {
                    bulletInstance = instantiate(this.bulletPrefab);
                }                
            }
            
            if(null != bulletInstance)
            {
                var bulletComponent = bulletInstance.getComponent( Maze_BulletBase.BulletBase );

                if(null != bulletComponent)
                {
                    bulletComponent.collisionGroup = this.bulletCollisionGroup;
                    bulletComponent.destroyBulletSubject.attach(this._destroyBulletObserver);
                }
            }
            else
            {
                throw("Error! bulletInstance == null!");
            }

            return bulletInstance;
        }

        protected destroyBullet(bullet:Node)
        {
            if(null != this.bulletPrefab)
            {
                var nodePool:NodePool|null = null;

                if(WeaponBase._bulletPoolMap.has(this.bulletPrefab.name))
                {
                    nodePool = WeaponBase._bulletPoolMap.get(this.bulletPrefab.name);
                }
                else
                {
                    nodePool = new NodePool();
                    WeaponBase._bulletPoolMap.set(this.bulletPrefab.name,nodePool);
                }

                var bulletComponent = bullet.getComponent( Maze_BulletBase.BulletBase );

                if(null != bulletComponent)
                {
                    bulletComponent.destroyBulletSubject.detach(this._destroyBulletObserver);
                }

                nodePool.put(bullet);               
            }   
        }

        // should be implemented by inhetired classes
        protected abstract actualFire() : void;

        private scheduleFire()
        {
            if(true == this._fireAllowed)
            {
                this.actualFire();
                this._fireAllowed = false;

                this.scheduleOnce(()=> 
                {
                    if(true == this._fireOn)
                    {
                        this.actualFire();
                    }

                    this._fireAllowed = true;
                    
                    if(true == this._fireOn)
                    {
                        this.scheduleFire();
                    }
                }, this.fireRate);
            }
        }

        public fireOn()
        {
            if(false == this._fireOn)
            {
                this._fireOn = true;
                this.scheduleFire();
            }
        }

        public fireOff()
        {
            if(true == this._fireOn)
            {
                this._fireOn = false;
            }
        }

        public start() 
        {
            this.node.addComponent(Maze_GlobalMouseListener.GlobalMouseListener);

            this.easyReference = new Maze_EasyReference.EasyReference(this.node);

            this._destroyBulletObserver.setObserverCallback((data:Maze_BulletBase.DestroyBulletContext) => 
            {                
                this.destroyBullet(data.bullet);
            });
        }
    }
}

File: Maze_Unity/Assets/Scripts/Weapon/WeaponBase.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using Maze_EasyReference;
using Maze_BulletBase;
using Maze_Observer;
using UnityEngine.Pool;
using UnityTimer;

namespace Maze_WeaponBase
{
    public abstract class WeaponBase : MonoBehaviour
    {
        private static Dictionary<string, ObjectPool<GameObject>> sBulletPoolMap = new Dictionary<string, ObjectPool<GameObject>>();

        [SerializeField]
        private GameObject mBulletPrefab;
        public GameObject BulletPrefab { get => mBulletPrefab; set => mBulletPrefab = value; }

        [SerializeField]
        private float mFireRate = 1;
        public float FireRate { get => mFireRate; set => mFireRate = value; }

        [SerializeField]
        private float mBulletTimeAlive = 2;
        public float BulletTimeAlive { get => mBulletTimeAlive; set => mBulletTimeAlive = value; }

        [SerializeField]
        private float mBulletSpeed = 2;
        public float BulletSpeed { get => mBulletSpeed; set => mBulletSpeed = value; }

        [SerializeField]
        private int mDamage = 5;
        public int Damage { get => mDamage; set => mDamage = value; }

        [SerializeField]
        private int mLayer = 0;
        public int Layer { get => mLayer; set => mLayer = value; }

        protected Maze_EasyReference.EasyReference mEasyReference;
        private bool mFireOn = false;

        private bool mFireAllowed = true;
        public bool FireAllowed { get => mFireAllowed; }

        private Vector3 mDirection = new Vector3();
        public Vector3 Direction { get => mDirection; set => mDirection = value; }

        private Maze_Observer.Observer<Maze_BulletBase.DestroyBulletContext>  mDestroyBulletObserver = 
            new Maze_Observer.Observer<Maze_BulletBase.DestroyBulletContext>();

        private UnityTimer.Timer mTimer = null;

        private ObjectPool<GameObject> findOrCreateBulletsPool(string name)
        {
            ObjectPool<GameObject> objectsPool = null;

            if (false == WeaponBase.sBulletPoolMap.TryGetValue(name, out objectsPool))
            {
                objectsPool = new ObjectPool<GameObject>(
                createFunc: () => { return Instantiate(mBulletPrefab); },
                actionOnGet: (GameObject obj) =>
                {
                    var bulletComponent = obj.GetComponent<Maze_BulletBase.BulletBase>();
                    if (null != bulletComponent)
                    {
                        bulletComponent.reuse();
                        bulletComponent.DestroyBulletSubject.attach(mDestroyBulletObserver);
                    }

                    obj.SetActive(true);
                },
                actionOnRelease: (GameObject obj) =>
                {
                    var bulletComponent = obj.GetComponent<Maze_BulletBase.BulletBase>();
                    if (null != bulletComponent)
                    {
                        bulletComponent.DestroyBulletSubject.detach(mDestroyBulletObserver);
                    }

                    obj.SetActive(false);
                });
                WeaponBase.sBulletPoolMap.Add(name, objectsPool);
            }

            return objectsPool;
        }

        protected GameObject createBullet()
        {
            GameObject bulletInstance = null;

            if (null != mBulletPrefab)
            {
                ObjectPool<GameObject> objectsPool = findOrCreateBulletsPool(mBulletPrefab.name);

                if (null != objectsPool)
                {
                    bulletInstance = objectsPool.Get();
                }
                else
                {
                    throw new System.Exception("[WeaponBase::createBullet] Error! objectsPool == null!");
                }

                if (null == bulletInstance)
                {
                    throw new System.Exception("[WeaponBase::createBullet] Error! bulletInstance == null!");
                }
            }

            return bulletInstance;
        }

        protected void destroyBullet(GameObject bullet)
        {
            if (null != mBulletPrefab)
            {
                ObjectPool<GameObject> objectsPool = findOrCreateBulletsPool(mBulletPrefab.name);

                if (null != objectsPool)
                {
                    objectsPool.Release(bullet);
                }
                else
                {
                    throw new System.Exception("[WeaponBase::createBullet] Error! objectsPool == null!");
                }
            }
        }

        protected abstract void actualFire();

        private void scheduleFire()
        {
            if (true == mFireAllowed)
            {
                actualFire();
                mFireAllowed = false;

                mTimer = UnityTimer.Timer.Register(mFireRate, () =>
                {
                    if (true == mFireOn)
                    {
                        actualFire();
                    }

                    mFireAllowed = true;

                    if (true == mFireOn)
                    {
                        scheduleFire();
                    }
                });
            }
        }

        public void fireOn()
        {
            if (false == mFireOn)
            {
                mFireOn = true;
                scheduleFire();
            }
        }

        public void fireOff()
        {
            if (true == mFireOn)
            {
                mFireOn = false;
            }
        }

        // Start is called before the first frame update
        void Start()
        {
            mEasyReference = new Maze_EasyReference.EasyReference(gameObject);

            mDestroyBulletObserver.setObserverCallback((Maze_BulletBase.DestroyBulletContext data) => 
            {
                destroyBullet(data.Bullet);
            });
        }

        // Update is called once per frame
        void Update()
        {

        }
    }
}